bessctl/pcapngdump.py

```diff
@@ -0,0 +1,192 @@
+# Copyright (c) 2017, Nefeli Networks, Inc.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+# list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+# this list of conditions and the following disclaimer in the documentation
+# and/or other materials provided with the distribution.
+#
+# * Neither the names of the copyright holders nor the names of their
+# contributors may be used to endorse or promote products derived from this
+# software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import binascii
+import struct
+
+# Section header block type
+SHB_TYPE = 0x0A0D0D0A
+# Enhanced packet block type
+EPB_TYPE = 0x00000006
+
+BYTE_ORDER_MAGIC = 0x1A2B3C4D
+# Recognized version number
+MAJ_VER = 1
+
+# Given the limited scope of this implementation, blocks bigger than this
+# are treated as errors.
+MAX_BLOCK_LEN = 0x1FFFF
+
+# Option codes
+OPT_CODE_END = 0x0000
+OPT_CODE_COMMENT = 0x0001
+
+BLOCK_COMMON_HEADER = 'II'
+BLOCK_COMMON_HEADER_SIZE = struct.calcsize(BLOCK_COMMON_HEADER)
+BLOCK_COMMON_FOOTER_SIZE = 4
+
+SH_BLOCK1 = 'IxxxxIxxxx'
+SH_BLOCK2 = 'xxxxIxxxxHH'
+SH_BLOCK_SIZE = struct.calcsize(SH_BLOCK1)
+
+EP_VALUE = 'xxxxxxxxxxxxII'
+EP_VALUE_SIZE = struct.calcsize(EP_VALUE)
+
+OPTION_HEADER = 'HH'
+OPTION_HEADER_SIZE = struct.calcsize(OPTION_HEADER)
+
+
+def read_exactly(f, size):
+    """
+    Read from file f exactly size bytes.
+    Might return a buffer smaller than size if EOF or error.
+    """
+    ret = b''
+    while True:
+        ret += f.read(size)
```

Comment:
Efficiency is probably not an issue here but if it is the usual trick is to read into a list of strings/bytestrings and then concatenate the list once at the end (to avoid O(n**2) appends)... it looks like size is generally <= 16 though, right?

Reply:
Ah good point. `size` actually depends on how big is a block and some of them are variable length.

I implemented your suggestion.

Also there was a bug in my implementation: on every iteration the code was subtracting the total length of ret, whiile it should have been subtracting only the length of the buffer returned on this iteration.

In practice read always returns what you ask, because this is not a tcp stream, but I wish python had a function for this (I'm not aware of one)

Thanks!